code/modules/reagents/chemistry/machinery/smoke_machine.dm

```diff
@@ -0,0 +1,138 @@
+	location = loc
+
+/obj/machinery/smoke_machine/Initialize()
+	. = ..()
+	create_reagents(volume)
+
+/obj/machinery/smoke_machine/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	return ..()
+
+/obj/machinery/smoke_machine/update_icon()
+	if((stat & NOPOWER) || (!on) || (reagents.total_volume == 0))
+		icon_state = "smoke0"
+	else
+		icon_state = "smoke1"
+	. = ..()
+
+/obj/machinery/smoke_machine/RefreshParts()
+	efficiency = 6
+	for(var/obj/item/stock_parts/matter_bin/B in component_parts)
+		efficiency += B.rating
+	for(var/obj/item/stock_parts/capacitor/C in component_parts)
+		efficiency += C.rating
+	for(var/obj/item/stock_parts/manipulator/M in component_parts)
+		efficiency += M.rating
+
+/obj/machinery/smoke_machine/process()
+	..()
+	update_icon()
+	if(stat & NOPOWER)
+		return
+	if(reagents.total_volume == 0)
+		on = FALSE
+		return
+	var/turf/T = get_turf(src)
+	var/smoke_test
+	smoke_test = locate(/obj/effect/particle_effect/smoke) in T.contents
+	if(on && (!smoke_test))
+		var/datum/effect_system/smoke_spread/chem/smoke_machine/smoke = new()
+		smoke.set_up(reagents, setting, efficiency, get_turf(src))
+		smoke.start()
+
+/obj/machinery/smoke_machine/attackby(obj/item/I, mob/user, params)
+	add_fingerprint(user)
+	if(istype(I, /obj/item/reagent_containers))
+		var/obj/item/reagent_containers/RC = I
+		var/units = RC.reagents.trans_to(src, RC.amount_per_transfer_from_this)
+		if(units)
+			to_chat(user, "<span class='notice'>You transfer [units] units of the solution to [src].</span>")
+			for(var/datum/reagent/R in RC.reagents.reagent_list)
+				add_logs(usr, src, "has added [R.name] to [src]")
+			return
+	if(default_unfasten_wrench(user, I))
+		return
+	return ..()
+
+/obj/machinery/smoke_machine/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, \
+										datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "smoke_machine", name, 450, 350, master_ui, state)
+		ui.open()
+
+/obj/machinery/smoke_machine/ui_data(mob/user)
+	var/data = list()
+	var/TankContents[0]
+	var/TankCurrentVolume = 0
+	if(reagents.reagent_list.len)
+		for(var/datum/reagent/R in reagents.reagent_list)
+			TankContents.Add(list(list("name" = R.name, "volume" = R.volume))) // list in a list because Byond merges the first list...
+			TankCurrentVolume += R.volume
+	data["TankContents"] = TankContents
+	data["isTankLoaded"] = reagents.total_volume ? 1 : 0
+	data["TankCurrentVolume"] = reagents.total_volume ? reagents.total_volume : null
+	data["TankMaxVolume"] = reagents.maximum_volume
+	data["active"] = on
+	data["setting"] = setting
+	data["screen"] = screen
+	data["analyzeVars"] = analyzeVars
+	return data
+
+/obj/machinery/smoke_machine/ui_act(action, params)
+	if(..() || !anchored)
+		return
+	switch(action)
+		if("purge")
+			reagents.clear_reagents()
+			. = TRUE
+
+		if("analyze")
```

Comment:
This whole section seems unused ? Analyze stuff is not anywhere in the template?

Reply:
I threw it in since the analyze functionality exists in other chemistry machines but format differences in the UI and design of these machines appears (this is the only chemistry machine that permanently stores chemicals and doesn't use a buffer) prevented me from getting this to work. I'm just going to cut it since analyze is mostly fluff anyway and it doesn't make sense for a smoke machine to have the same analysis functionality as a chem dispenser.

Reply:
yeah remove it please

Reply:
@Cyberboss